darts/utils/likelihood_models.py

```diff
@@ -35,64 +153,852 @@ def _sample(self, model_output: torch.Tensor) -> torch.Tensor:
-    """
+class PoissonLikelihood(Likelihood):
+    def __init__(self, prior_lambda=None, prior_strength=1.):
+        """
+        Poisson distribution. Can typically be used to model event counts during time intervals, when the events
+        happen independently of the time since the last event.
+
+        https://en.wikipedia.org/wiki/Poisson_distribution
+
+        - Univariate discrete distribution
+        - Support: :math:`\mathbb{N}_0` (natural numbers including 0).
+        - Parameter: rate :math:`\\lambda > 0`.
+
+        Parameters
+        ----------
+        prior_lambda
+            rate :math:`\\lambda` of the prior Poisson distribution (default: None)
+        prior_strength
+            strength of the loss regularisation induced by the prior
+        """
+        self.prior_lambda = prior_lambda
+        _check_strict_positive(self.prior_lambda, 'lambda')
+
+        self.nllloss = nn.PoissonNLLLoss(log_input=False, full=True)
+        self.softplus = nn.Softplus()
+        super().__init__(prior_strength)
 
+    def _nllloss(self, params_out, target):
+        lambda_out = params_out
+        return self.nllloss(lambda_out, target)
+
+    @property
+    def _prior_params(self):
+        return self.prior_lambda,
+
+    def _distr_from_params(self, params):
+        lmbda = params[0]
+        return _Poisson(lmbda)
+
+    def sample(self, model_output: torch.Tensor) -> torch.Tensor:
+        model_lambda = self._params_from_output(model_output)
+        return torch.poisson(model_lambda)
+
+    @property
+    def num_parameters(self) -> int:
+        return 1
+
+    def _params_from_output(self, model_output):
+        lmbda = self.softplus(model_output)
+        return lmbda
+
+
+class NegativeBinomialLikelihood(Likelihood):
     def __init__(self):
-        self.loss = nn.PoissonNLLLoss(log_input=False)
+        """
+        Negative Binomial distribution.
+
+        https://en.wikipedia.org/wiki/Negative_binomial_distribution
+
+        It does not support priors.
+
+        - Univariate discrete distribution.
+        - Support: :math:`\mathbb{N}_0` (natural numbers including 0).
+        - Parameters: number of failures :math:`r > 0`, success probability :math:`p \in (0, 1)`.
+
+        Behind the scenes the distribution is reparameterized so that the actual outputs of the
+        network are in terms of the mean :math:`\\mu` and shape :math:`\\alpha`.
+        """
         self.softplus = nn.Softplus()
         super().__init__()
 
-    def _compute_loss(self, model_output: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
-        model_output = self._lambda_from_output(model_output)
-        return self.loss(model_output, target)
+    @property
+    def _prior_params(self):
+        return None
+
+    @staticmethod
+    def _get_r_and_p_from_mu_and_alpha(mu, alpha):
+        # See https://en.wikipedia.org/wiki/Negative_binomial_distribution for the different parametrizations
+        r = 1. / alpha
```

Comment:
I think this is causing some numerical issues. `r` can end up having `inf` values here, which then results in NaNs for `p` too. I changed this line to `r = 1. / (alpha + eps)` and get good results with eps=1e-3, but not with other values. Obviously we should find a solution without magic numbers, but I don't have one yet.

Reply:
Good one. Actually I think the `eps` trick is fine, that's a standard way to approach it. However we could maybe find a smaller value? How about something like 1e-6?